File: app/src/main/java/dk/itu/garbage/ItemsDB.java
package dk.itu.garbage;

import android.content.Context;

import androidx.lifecycle.ViewModel;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;

//now using ViewModel
public class ItemsDB extends ViewModel {
    //a static factory method with the return type as an object of this singleton class
    private static ItemsDB sItemsDB;
    private final Map<String, String> itemsMap = new HashMap<>();

    //declaring access modifier of constructor private /changed to use context with AndroidViewModel
    protected ItemsDB(Context context) {
        fillItemsDB(context, "garbage.txt");
    }



    public static ItemsDB get() {
        if (sItemsDB == null) {
            throw new IllegalStateException("ItemsDB must be initialized");
        }
        return sItemsDB;
    }

    //other methods: getWhere

    public Map<String, String> getItemsDBMap() {
        return itemsMap;
    }


    // takes input from TextEdit and searches Item object for equality
    public String searchItems(String input) {
        String dbItem;
        if (itemsMap.containsKey(input.toLowerCase().trim())) {
           dbItem = input + " should be placed in: " + itemsMap.get(input);
           return dbItem;
        }
        return input + " not found";
    }

    //see also ItemsModelView
    public void removeItem(String delItem) {
        for (String i : itemsMap.keySet()) {
            if (delItem.equals(i)) {
                itemsMap.remove(i);
                break; //necessary?
            }
        }
    }

    public int getSize() {
        return itemsMap.size();
    }

    //see also ItemsModelView
    public void addItem(String what, String where) {
        itemsMap.put(what, where);
    }

    /** The map of the ItemsDB, should be filled with the category/items from file in assets folder */
    private void fillItemsDB(Context context, String filename) {
        //get items and categories from file
        try {
            //read file
            BufferedReader reader = new BufferedReader(new InputStreamReader(context.getAssets().open(filename)));

            //assign the return value of readLine to variable.
            String line = reader.readLine();
            //while there is something in the file
            while (line != null) {
                // assign input to array, split at white space
                String[] itemsFile = line.toLowerCase().split(",");
                // put the the input into the map ofItemsDB as k/v pair
                itemsMap.put(itemsFile[0], itemsFile[1]);
                //get content of next line
                line = reader.readLine();
            }

        } catch (IOException e) {
            System.out.println(e);
        }
    }

    public String listItems() {
        String dbItems = "";
        for (Map.Entry item : itemsMap.entrySet()) {
            dbItems = dbItems + item.getKey() + " should be placed in: " + item.getValue() + "\n";
        }
        return dbItems;
    }

}

File: app/src/main/java/dk/itu/garbage/ItemsDB.kt
package dk.itu.garbage

import android.content.Context
import java.util.HashMap
import java.io.BufferedReader
import java.io.IOException

class ItemsDB(context: Context) {
    // val is equivalent to finale
    private val itemsMap : HashMap<String, String> = HashMap<String, String>()

    init {
        fillItemsDB(context, "garbage.txt")
    }

    companion object {
        private var sItemsDB : ItemsDB? = null

        fun initialize(context: Context) {
            if (sItemsDB == null) {
                sItemsDB = ItemsDB(context)
            }
        }

        fun get() : ItemsDB {
            return sItemsDB ?:
            throw IllegalStateException("ItemsDB must be initialized")
        }
    }

    fun size() : Int {
        return itemsMap.size
    }

    fun searchItems(input : String) : String {
        val dbItem: String

        if (itemsMap.containsKey(input)) {
            // use indexing with map
            dbItem = "$input should be placed in: ${itemsMap[input]}"
            return dbItem
        }
        return "$input not found"
    }

    fun addItem(what : String, where : String) {
        itemsMap[what] = where
    }

    private fun fillItemsDB(context: Context, filename : String) {
            //get items and categories from file
            try {
                //read file
                var reader : BufferedReader = context.assets.open(filename).bufferedReader()
                //assign the return value of readLine to variable.
                var line = reader.readLine();
                //while there is something in the file
                while (line != null) {
                    // assign input to array, split at white space
                    val itemsFile = line.lowercase().split(",");
                    // put the the input into the map ofItemsDB as k/v pair
                    itemsMap[itemsFile[0]] = itemsFile[1]
                    //get content of next line
                    line = reader.readLine();
                }
            } catch (e : IOException) {
                println(e)
            }
        }



        /*
        val reader : BufferedReader = File("src/main/assets/garbage.txt").bufferedReader()
        val lineList = mutableListOf<String>()
        reader.useLines {
            lines -> lines.forEach {
                lineList.add(it)
            }
            lineList.forEach {
                val line = it.split(",")
                itemsMap[line[0]] = line[1]
            }
        }
        */

}


